CAMEL-10577 replace BufferingResponseListener with InputStreamRespons…

kevinearls opened the pull request:
…eListener to avoid hardcoded 2MB limit

oscerd commented on components/camel-jetty9/src/main/java/org/apache/camel/component/jetty9/JettyContentExchange9.java:
```diff
@@ -214,14 +220,28 @@ public void onFailure(Request request, Throwable failure) {
                     doTaskCompleted(result.getFailure());
                 } else {
-                    onResponseComplete(result, getContent(), getMediaType());
+                    onResponseComplete(result, baos.toByteArray(), null);
```
Is there a reason for null as media Type? Just curiosity :-)

Thanks, I will test and merge

kevinearls replied:
onResponseComplete takes mediaType as a parameter, but does not use it.  BufferingResponseListener has a getMediaType method, but InputStreamResponseListener.  I could just remove it from onResponseComplete instead.

davsclaus commented on components/camel-jetty9/src/main/java/org/apache/camel/component/jetty9/JettyContentExchange9.java:
```diff
@@ -214,14 +220,28 @@ public void onFailure(Request request, Throwable failure) {
+
+            @Override
+            public void onContent(Response response, ByteBuffer content, Callback callback) {
+                if (response.getStatus() != SC_OK) {
```
Why this WARN log, wont you be able to get some error and still get data. I would remove that WARN

kevinearls replied:
I wasn't sure what to do in the case of a non-200 status here.  Treating this as a failure (in this case calling doTaskCompleted()) caused a bunch of other tests to fail.  Should I treat this as a failure and update the other tests, or just ignore it?

davsclaus replied:
How do you know that the response is finished? What if there is a lot of data, wont this onContent be called multiple times?

kevinearls replied:
When the response is finished onComplete is called.  I don't think there's anyway to know in onContent if you've received all the data.

davsclaus replied:
Ah okay but a byte array stream is unfortunately hard limited, so if onContent is called multiple times the buffer looses data etc. What is needed is to be able to store the data in a growing buffer.

You can try to do like the MarshalProcessor which uses Camel's stream cache to write to its buffer. Then when you are done, you can get the result from its build method.

davsclaus commented on components/camel-jetty9/src/main/java/org/apache/camel/component/jetty9/JettyContentExchange9.java:
```diff
@@ -214,14 +220,28 @@ public void onFailure(Request request, Throwable failure) {
+                if (response.getStatus() != SC_OK) {
+                    LOG.warn("Response received {}: {}", response.getStatus(), response.getReason());
+                }
+                byte[] buffer = new byte[content.limit()];
```
This will stream all the data into memory which is not what we want. Either we should use Camel's stream caching or some other kind of solution as people want to stream bigger data from jetty without having to have all the data in memory.

Camel has a stream caching that can overflow to disk for big data. Which we use in data formats etc and other places.

I wonder what others are doing with Jetty and this kind of listener.

davsclaus commented:
However if the old code has a hardcoded 2mb limit anyway then this change is still better because it wont have that limit. Then we can log another JIRA ticket to look into a way to stream the data without an in memory byte[] buffer.

@kevinearls can you log a JIRA and then polish this PR.

kevinearls commented:
I've removed the unused contentType parameter from onResponseComplete as @oscerd suggested, and removed the warning as @davsclaus suggested.  I made an initial attempt to replace the ByteArrayOutputStream with an OutputStreamBuilder, but that resulted in the JettyStreamCacheIssueTest hanging.  

I'll open a new JIRA to cover that.

davsclaus commented:
@oscerd looks good to me now. Feel free to merge to master and 2.18.x branch (if possible)

oscerd commented:
Thanks, the PR has been merged. Can you close the PR?

oscerd commented:
Merged on master and 2.18.x

davsclaus commented:
Thanks for the PR it has been merged. Do you mind closing this?
